Document how to pack the ZODB when using containers (#1886)
plone/documentation: docs/install/containers/recipes/index.md

[handler_eventlog]
class = StreamHandler
args = (sys.stderr,)
level = INFO
formatter = generic
```

Comparing this file with the [original `zope.ini` file](https://github.com/plone/plone-backend/blob/6.1.x/skeleton/etc/zope.ini) that comes with the `plone/plone-backend` container, you may realize that the only change is the `translogger` configuration was removed from the `pipeline` section.
This [`translogger` middleware produces logs in the Apache Combined Log Format](https://docs.pylonsproject.org/projects/waitress/en/latest/logging.html).
The above configuration removes it from the setup.

After adding the {file}`zope.ini` file in your project, adjust the {file}`Dockerfile` by inserting the command `COPY zope.ini etc/` before the `RUN` command as highlighted below.
This new command copies the {file}`zope.ini` file into the container.

```{code-block} dockerfile
:emphasize-lines: 4

# Add local code
COPY scripts/ scripts/
COPY . src
COPY zope.ini etc/

# Install local requirements and pre-compile mo files
RUN <<EOT
```

After making these changes, build the project container as usual.
It will no longer output the access log, but will continue to output the event log.

> ## Pack the ZODB

Reviewer: @erral @davisagli would this section be more appropriate under the existing [Advanced usage in `plone/plone-backend`](https://6.docs.plone.org/install/containers/images/backend.html#advanced-usage)?

After making these changes, build the project container as usual.
It will no longer output the access log, but will continue to output the event log.

## Pack the ZODB

> A common maintenance task of a Plone instance is to pack the ZODB.

Reviewer: ```suggestion
A common maintenance task of a Plone instance is to [pack the ZODB](https://zodb.org/en/stable/reference/zodb.html#ZODB.DB.pack).
Packing removes old revisions of objects.
It is similar to [routine vacuuming in PostgreSQL](https://www.postgresql.org/docs/8.3/routine-vacuuming.html).
```

> The official `plone/plone-backend` container and project containers based on them have a `pack` command to pack the ZODB.

Reviewer: If this is moved into https://6.docs.plone.org/install/containers/images/backend.html, then ignore this suggestion.

```suggestion
The official {doc}`/install/containers/images/backend` container and project containers based on them have a `pack` command to pack the ZODB.
```

> 

Reviewer: ```suggestion
```

> The command will work on standalone mode, ZEO mode and Relstorage mode (with Posgres).

Reviewer: ```suggestion
The command will work in standalone mode, ZEO mode, and RelStorage mode but only with PostgreSQL.
```

> To run the command you have to run your container passing the command:

Reviewer: ```suggestion
Invoke the command in a running container by passing in the appropriate command for the mode.
```

> In standalone mode you will have a volume mounting the ZODB, so it will be something like:

Reviewer: ```suggestion
In standalone mode, ZODB is in a mounted volume, so the command would be similar to the following.
```

> ```

Reviewer: ````suggestion
```shell
````

docker run -v /path/to/your/volume:/data plone/plone-backend pack
```

> In ZEO mode, you will have to run it next to your ZEO instance:

Reviewer: ```suggestion
In ZEO mode, run the command next to your ZEO instance.
```

> ```

Reviewer: ````suggestion
```shell
````

docker run -e ZEO_ADDRESS=zeo:8100 --link zeo plone/plone-backend pack
```

> In Relstorage, you will need to pass the connection DSN:

Reviewer: ```suggestion
In RelStorage mode, pass the connection DSN.
```

> ```

Reviewer: ````suggestion
```shell
````

docker run -e RELSTORAGE_DSN="dbname='plone' user='plone' host='db' password='password' port='5432'" pack
```

> If you are running your containers using docker compose, the command is much easier:

Reviewer: ```suggestion
In running containers that use Docker Compose, the command is less complicated.
```

> ```

Reviewer: ````suggestion
```shell
````

docker compose run backend pack
```

> Provided that the name of the service that runs the Plone instance is `backend`, otherwise exchange `backend` with your container's name.

Reviewer: ```suggestion
The above command assumes that the service that runs the Plone instance is named `backend`.
Otherwise replace `backend` with your container's name.
```

After making these changes, build the project container as usual.
It will no longer output the access log, but will continue to output the event log.

## Pack the ZODB

A common maintenance task of a Plone instance is to [pack the ZODB](https://zodb.org/en/stable/reference/zodb.html#ZODB.DB.pack).
Packing removes old revisions of objects.
It is similar to [routine vacuuming in PostgreSQL](https://www.postgresql.org/docs/8.3/routine-vacuuming.html).

> The official {doc}`/install/containers/images/backend` container and project containers based on them have a `pack` command to pack the ZODB.

Reviewer 2: This should have a `versionadded` note. I just merged the PR which adds this command today, and it won't actually be available in the images until the next Plone releases are made. The backport PR for Plone 6.0 is still pending.

> The command will work in standalone mode, ZEO mode, and RelStorage mode but only with PostgreSQL.

Reviewer 2: This sentence is confusing. "but only with PostgreSQL" is meant to be a qualifier only when using it with Relstorage, but it sounds like it applies to all the options, which makes no sense (it's only _possible_ to use postgres if you're doing via Relstorage).

Invoke the command in a running container by passing in the appropriate command for the mode.

In standalone mode, ZODB is in a mounted volume, so the command would be similar to the following.

```shell
docker run -v /path/to/your/volume:/data plone/plone-backend pack
```

In ZEO mode, run the command next to your ZEO instance.

```shell
docker run -e ZEO_ADDRESS=zeo:8100 --link zeo plone/plone-backend pack
```

In RelStorage mode, pass the connection DSN.

```shell
docker run -e RELSTORAGE_DSN="dbname='plone' user='plone' host='db' password='password' port='5432'" pack

> ```

Reviewer 2: In Docker Swarm, it's also necessary to specify the network if the database is only available on an internal network. It might be easier to use `docker exec` to run the pack command in an existing plone-backend container.
